src/topology/order/scott.lean

```diff
@@ -0,0 +1,469 @@
+- `directed_lub_mem_implies_tail_subset` - the topological space where a set `u` is open if, when
+  the least upper bound of a directed set `d` lies in `u` then there is a tail of `d` which is a
+  subset of `u`.
+- `with_scott_topology.topological_space` - the Scott topology is defined as the join of the
+  `directed_lub_mem_implies_tail_subset` topology and the topology of upper sets.
+
+## Main statements
+
+- `with_scott_topology.is_open_is_upper` - Scott open sets are upper
+- `with_scott_topology.is_closed_is_lower` - Scott closed sets are lower
+- `with_scott_topology.continuous_monotone` - Scott continuous functions are monotone.
+- `preserve_lub_on_directed_iff_scott_continuity` - a function preserves least upper bounds of
+  directed sets if and only if it is Scott continuous
+- `with_scott_topology.t0_space` - the Scott topology on a partial order is T₀
+
+## Implementation notes
+
+A type synonym `with_scott_topology` is introduced and for a preorder `α`, `with_scott_topology α`
+is made an instance of `topological_space` by the topology generated by the complements of the
+closed intervals to infinity.
+
+A class `Scott` is defined in `topology.omega_complete_partial_order` and made an instance of a
+topological space by defining the open sets to be those which have characteristic functions which
+are monotone and preserve limits of countable chains. Whilst this definition of the Scott topology
+coincides with the one given here in some special cases, in general they are not the same
+[Domain Theory, 2.2.4][abramsky_gabbay_maibaum_1994].
+
+## References
+
+* [Gierz et al, *A Compendium of Continuous Lattices*][GierzEtAl1980]
+* [Abramsky and Jung, *Domain Theory*][abramsky_gabbay_maibaum_1994]
+
+## Tags
+
+Scott topology, preorder
+
+-/
+
+variables (α β : Type*)
+
+open set
+
+section preorder
+
+variables {α} {β}
+
+variables [preorder α] [preorder β]
+
+lemma is_upper_set_iff_forall_le  {s : set α} : is_upper_set s ↔ ∀ ⦃a b : α⦄, a ≤ b →
+  a ∈ s → b ∈ s := iff.rfl
+
+/--
+The set of upper sets forms a topology
+-/
+def upper_set_topology : topological_space α :=
+{ is_open := is_upper_set,
+  is_open_univ := is_upper_set_univ,
+  is_open_inter := λ _ _, is_upper_set.inter,
+  is_open_sUnion := λ _, is_upper_set_sUnion }
+
+/--
+The set of sets satisfying "property (S)" ([GierzEtAl1980] p100) form a topology
+-/
+def directed_lub_mem_implies_tail_subset : topological_space α :=
+{ is_open := λ u, ∀ (d : set α) (a : α), d.nonempty → directed_on (≤) d → is_lub d a → a ∈ u →
+               ∃ b ∈ d, (Ici b)∩ d ⊆ u,
+  is_open_univ := begin
+    intros d a hd₁ hd₂ hd₃ ha,
+    cases hd₁ with b hb,
+    use b,
+    split,
+    { exact hb, },
+    { exact (Ici b ∩ d).subset_univ, },
+  end,
+  is_open_inter := begin
+    rintros s t,
+    intros hs,
+    intro ht,
+    intros d a hd₁ hd₂ hd₃ ha,
+    cases (hs d a hd₁ hd₂ hd₃ ha.1) with b₁ hb₁,
+    cases (ht d a hd₁ hd₂ hd₃ ha.2) with b₂ hb₂,
+    cases hb₁,
+    cases hb₂,
```

Comment:
I would imagine you can compress this into two `obtain `lines

Reply:
Thanks, this is true in several other places as well.